### ISSUE TITLE
Concurrent map writes kills dgraph server

### ISSUE THREAD

**Issue by chilts** [dgraph-io/dgraph#1543]
When trying to insert all of the Enron emails and the graph between emails and people, I managed to kill the dgraph server. When I'm loading say 10,000 emails, I'll be able to replicate a crash about 1 in 2 times, however it doesn't necessarily occur at the same stage.

This leads me to believe it is a race condition. Please see this error from my logs, and I can include the entire log if you wish.

I'm running `dgraphzero` locally with:

```
$ dgraphzero -id 1 2>&1 | tee ../dgraphzero.log
```

And running `dgraph` with:

```
$ dgraph --memory_mb 4096 --peer :8888 2>&1 | tee ../dgraph.log
```

Here's the part of the error I think is the culprit:

```
fatal error: concurrent map writes

goroutine 30975 [running]:
runtime.throw(0xdb17fb, 0x15)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/panic.go:605 +0x95 fp=0xc42a7f2748 sp=0xc42a7f2728 pc=0x42f385
runtime.mapassign_faststr(0xcb9c60, 0xc43497d3e0, 0xc432e88004, 0x5, 0x14e4e20)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/hashmap_fast.go:685 +0x4da fp=0xc42a7f27c8 sp=0xc42a7f2748 pc=0x4107aa
github.com/dgraph-io/dgraph/query.(*SubGraph).populateUidValVar(0xc42fece780, 0xc43497d3e0, 0x14e3f08, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:1450 +0xb9e fp=0xc42a7f2aa0 sp=0xc42a7f27c8 pc=0xbca6be
github.com/dgraph-io/dgraph/query.(*SubGraph).assignVars(0xc42fece780, 0xc43497d3e0, 0xc42a7f2b48, 0x0, 0x0, 0x0, 0x2)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:1414 +0x106 fp=0xc42a7f2b08 sp=0xc42a7f2aa0 pc=0xbc9a66
github.com/dgraph-io/dgraph/query.ProcessGraph(0x144b4a0, 0xc432972300, 0xc42fece780, 0x0, 0xc432972720)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:1919 +0xb79 fp=0xc42a7f3030 sp=0xc42a7f2b08 pc=0xbcdd29
github.com/dgraph-io/dgraph/query.(*SubGraph).upsert(0xc42fece780, 0x144b4a0, 0xc432972300, 0x0, 0xc42fecec00, 0x0)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:2383 +0x364 fp=0xc42a7f3170 sp=0xc42a7f3030 pc=0xbd23d4
github.com/dgraph-io/dgraph/query.(*QueryRequest).ProcessQuery(0xc42a7f3c00, 0x144b4a0, 0xc432972300, 0xc432c5a000, 0x0, 0xa)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:2524 +0xc4d fp=0xc42a7f3378 sp=0xc42a7f3170 pc=0xbd325d
github.com/dgraph-io/dgraph/query.(*QueryRequest).ProcessWithMutation(0xc42a7f3c00, 0x144b4a0, 0xc432972300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc43497d380, ...)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:2682 +0x6f9 fp=0xc42a7f37f8 sp=0xc42a7f3378 pc=0xbd49d9
main.queryHandler(0x144a7a0, 0xc42fe3c0e0, 0xc42fa94700)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/cmd/dgraph/main.go:299 +0x90b fp=0xc42a7f3cb0 sp=0xc42a7f37f8 pc=0xbe4b9b
net/http.HandlerFunc.ServeHTTP(0xdda6b0, 0x144a7a0, 0xc42fe3c0e0, 0xc42fa94700)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:1918 +0x44 fp=0xc42a7f3cd8 sp=0xc42a7f3cb0 pc=0x69e824
net/http.(*ServeMux).ServeHTTP(0x14c3ac0, 0x144a7a0, 0xc42fe3c0e0, 0xc42fa94700)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2254 +0x130 fp=0xc42a7f3d18 sp=0xc42a7f3cd8 pc=0x69fed0
net/http.serverHandler.ServeHTTP(0xc42a1a8270, 0x144a7a0, 0xc42fe3c0e0, 0xc42fa94700)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2619 +0xb4 fp=0xc42a7f3d48 sp=0xc42a7f3d18 pc=0x6a1444
net/http.(*conn).serve(0xc42c471540, 0x144b420, 0xc43281ea40)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:1801 +0x71d fp=0xc42a7f3fc8 sp=0xc42a7f3d48 pc=0x69d6fd
runtime.goexit()
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc42a7f3fd0 sp=0xc42a7f3fc8 pc=0x460381
created by net/http.(*Server).Serve
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2720 +0x288
```

In file `query/query.go` I can see the following code:

```
	} else if len(sg.DestUIDs.Uids) != 0 {
		// This implies it is a entity variable.
		if v, ok = doneVars[sg.Params.Var]; !ok {
			doneVars[sg.Params.Var] = varValue{
				Uids: sg.DestUIDs,
				path: sgPath,
			}
			return nil
		}
		// For a recurse query this can happen. We don't allow using the same variable more than
		// once otherwise.
		uids := v.Uids
		lists := make([]*protos.List, 0, 2)
		lists = append(lists, uids, sg.DestUIDs)
		v.Uids = algo.MergeSorted(lists)
		doneVars[sg.Params.Var] = v
```

So I presume concurrent access to `doneVars[]` is making it fail. This is the current 8.2-dev release. I was also able to do this using last week's 8.2 release. I noted the code had changed since then which is why I wanted to try 8.2-dev. Last week's code in 8.2 was just the following but with the same fatal effect.

```
	} else if len(sg.DestUIDs.Uids) != 0 {
		// This implies it is a entity variable.
		doneVars[sg.Params.Var] = varValue{
			Uids: sg.DestUIDs,
			path: sgPath,
		}
```

I apologise ... my plan was to try and put a Mutex in there (just to stop it happening) and see if I could run through the entire set of emails but I've had trouble building and running dgraph locally due to other reasons which I'll raise as a separate issue.

As further information for this bug, I think it's pertinent to see an example of what I'm putting into dgraph. The reason I say this is because this is related to the new `upsert` functionality and whilst your examples on the site always show just one xid being upserted, I'm sometimes doing 10, 20 or 30 upserts with the same mutation so I wonder if that is causing the issues (and/or whether that particular use-case has been tested as much).

Here's one example of a document I'm sending (via a plain http POST in Go, see later).

```
{
  email as var(func: eq(xid, "<2519643.1075860072192.JavaMail.evans@thyme>")) @upsert
  human as var(func: eq(xid, "mark.taylor@enron.com")) @upsert
  human_to_0 as var(func: eq(xid, "leonardo.pacheco@enron.com")) @upsert
  human_cc_0 as var(func: eq(xid, "jay.hawthorn@enron.com")) @upsert
  human_cc_1 as var(func: eq(xid, "cynthia.harkness@enron.com")) @upsert
  human_cc_2 as var(func: eq(xid, "jean.mrha@enron.com")) @upsert
  human_cc_3 as var(func: eq(xid, "david.forster@enron.com")) @upsert
  human_cc_4 as var(func: eq(xid, "julie.ferrara@enron.com")) @upsert
  human_cc_5 as var(func: eq(xid, "kal.shah@enron.com")) @upsert
  human_bcc_0 as var(func: eq(xid, "jay.hawthorn@enron.com")) @upsert
  human_bcc_1 as var(func: eq(xid, "cynthia.harkness@enron.com")) @upsert
  human_bcc_2 as var(func: eq(xid, "jean.mrha@enron.com")) @upsert
  human_bcc_3 as var(func: eq(xid, "david.forster@enron.com")) @upsert
  human_bcc_4 as var(func: eq(xid, "julie.ferrara@enron.com")) @upsert
  human_bcc_5 as var(func: eq(xid, "kal.shah@enron.com")) @upsert
}
mutation {
  set {
   uid(email) <xid> "<2519643.1075860072192.JavaMail.evans@thyme>" .
   uid(email) <filename> "data/maildir/taylor-m/sent/1018." .
   uid(email) <date> "2000-05-01 01:31:00 -0700 PDT" .
   uid(human) <sent> uid(email) .
   uid(email) <from> "mark.taylor@enron.com" .
   uid(email) <subject> "Re: Bandwidth Launch on EOL Website Ticker Text" .

   uid(human_to_0) <xid> "leonardo.pacheco@enron.com" .
   uid(human_cc_0) <xid> "jay.hawthorn@enron.com" .
   uid(human_cc_1) <xid> "cynthia.harkness@enron.com" .
   uid(human_cc_2) <xid> "jean.mrha@enron.com" .
   uid(human_cc_3) <xid> "david.forster@enron.com" .
   uid(human_cc_4) <xid> "julie.ferrara@enron.com" .
   uid(human_cc_5) <xid> "kal.shah@enron.com" .
   uid(human_bcc_0) <xid> "jay.hawthorn@enron.com" .
   uid(human_bcc_1) <xid> "cynthia.harkness@enron.com" .
   uid(human_bcc_2) <xid> "jean.mrha@enron.com" .
   uid(human_bcc_3) <xid> "david.forster@enron.com" .
   uid(human_bcc_4) <xid> "julie.ferrara@enron.com" .
   uid(human_bcc_5) <xid> "kal.shah@enron.com" .
   uid(human) <xid> "mark.taylor@enron.com" .
   uid(human) <sent_to> uid(human_to_0) .
   uid(human_to_0) <received_from> uid(human) .
   uid(email) <to> uid(human_to_0) .
   uid(human_to_0) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_0) .
   uid(human_cc_0) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_0) .
   uid(human_cc_0) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_1) .
   uid(human_cc_1) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_1) .
   uid(human_cc_1) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_2) .
   uid(human_cc_2) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_2) .
   uid(human_cc_2) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_3) .
   uid(human_cc_3) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_3) .
   uid(human_cc_3) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_4) .
   uid(human_cc_4) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_4) .
   uid(human_cc_4) <received> uid(email) .
   uid(human) <sent_cc> uid(human_cc_5) .
   uid(human_cc_5) <received_from> uid(human) .
   uid(email) <cc> uid(human_cc_5) .
   uid(human_cc_5) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_0) .
   uid(human_bcc_0) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_0) .
   uid(human_bcc_0) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_1) .
   uid(human_bcc_1) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_1) .
   uid(human_bcc_1) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_2) .
   uid(human_bcc_2) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_2) .
   uid(human_bcc_2) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_3) .
   uid(human_bcc_3) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_3) .
   uid(human_bcc_3) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_4) .
   uid(human_bcc_4) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_4) .
   uid(human_bcc_4) <received> uid(email) .
   uid(human) <sent_bcc> uid(human_bcc_5) .
   uid(human_bcc_5) <received_from> uid(human) .
   uid(email) <bcc> uid(human_bcc_5) .
   uid(human_bcc_5) <received> uid(email) .
  }
}
```

As I write this I realise that since `jay.hawthorn@enron.com` is in both the Cc and the Bcc lists, if I'd need to de-duplicate within the upsert section and map `human_bcc_0` to `human_cc_0`. Up until now I didn't think this mattered, but perhaps it does. I have no evidence whether the server crashes when I have duplicates or not, but I can try and find out for you if you think it's important. Or you might be able to tell me if I should try without duplicates in this block. I'd suggest this **should** work, but you'll be able to tell me otherwise. :)

And for completeness, here's my function for sending a string, the string being what is shown above:

```
func sendToDgraph(query []byte) error {
	fmt.Println("sendToDgraph()")

	url := "http://localhost:8080/query"
	fmt.Println("URL:>", url)

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(query))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	fmt.Println("response Status:", resp.Status)
	fmt.Println("response Headers:", resp.Header)
	body, _ := ioutil.ReadAll(resp.Body)
	fmt.Println("response Body:", string(body))

	return nil
}
```

Many thanks for dgraph, I've had a lot of fun playing with it this past week. :+1: :100: :clap:

**Comment by manishrjain** [comment]
Thanks a lot, @chilts for the detailed issue. And thanks for trying Dgraph out. Glad you liked it!
@pawanrawal worked on the upsert functionality, so I'll assign it to him

**Comment by pawanrawal** [comment]
Thanks for the detailed bug report @chilts. I was able to reproduce the bug and figure out the cause very easily because of your report.

I have a fix in place which will be merged to release and master after it is approved.

**Comment by chilts** [comment]
Oh wow thanks @pawanrawal. I'll try this again with the nightly release once the PR is merged. (I'm assuming it'll get automatically included, is that right?)

**Comment by pawanrawal** [comment]
That's right @chilts. In fact, even the release binary for v0.8.2 would be updated as we want all new users to get the bugfix.

**Comment by pawanrawal** [comment]
That's right @chilts. We don't create a new git tag here but just update the released binaries so that new users can get it. 

Let us know how does it go for you with the new binaries.
